docs/source/graph.rst

```diff
@@ -14,4 +14,27 @@ Base class for running neural networks in Static Graph Mode.
             name,
             debug,
             __repr__,
+            _add_block,
```

Comment:
顺便一提，`__repr__` 这个历史遗留问题是不是也可以不用导出了，这也是个特殊的私有方法……

Reply:
repr是想告诉用户，该对象可以打印

Reply:
我觉得，像`__`这类方法，几乎都没意向让用户显式调用，也就没有必要让用户显式感知。比如这个repr，是不是不导出。然后在Graph的docstring下感知graph对象可以print，并给个example，就可以了

Reply:
其实torch里也出现了__init__和__config__这样类似的方法的文档。

Reply:
但是没有 `__repr__` 吧，我昨天搜是一个都没有。 init那个比较特殊，是因为会当做类的初始化构造的文档

config我确实不了解

Reply:
想了下还是应该导出，比如现在导出了`__call__`和`__repr__`方法。

它们都可以显示调用，不过是因为python提供了语法糖：
- `__call__``可以被用`()`调用
- `print(obj)`会调用`print(obj.__repr__())`

`__call__`和`__repr__`实际含义是，该类实现了用于调用和打印的自定义操作符。

属于比较标准的做法。而且显示它们存在，用户才知道这个自定义的类可以用它们，以及它们的用法。

> A class can implement certain operations that are invoked by special syntax (such as arithmetic operations or subscripting and slicing) by defining methods with special names. This is Python’s approach to operator overloading, allowing classes to define their own behavior with respect to language operators. 

https://docs.python.org/3/reference/datamodel.html#special-method-names

Reply:
总结一下，就是这些自定义方法需要让用户知道，只是python提供了语法糖来调用这些方法。

nn.Graph的这些方法的调用 or 调用结果带有复杂性。
- `__call__` 的输入输出，编译时间，至少要像用户展示该方法存在及其用法
- `__repr__` 的结构并不naive，需要解释，至少要像用户展示该方法存在及其用法

所以他们还是需要在API Doc去说明的，要不没有直接承载它们说明的地方。

Reply:
> 想了下还是应该导出，比如现在导出了 `__call__` 和 `__repr__` 方法。

这个我的看法不一样。他们虽然 “可以显示调用”，但是其实 python 恰恰不希望用户显式调用，而是希望用户用语法糖的方式去调用，也就是运算符重载的 `xxx()` 和 `print(xxx)`。

而这类 https://docs.python.org/3/reference/datamodel.html#special-method-names 信息，是 Python 的知识，放在 OneFlow 中其实也不恰当（我们应该预设用户知道这些内容，或者不知道这些内容也不影响它操作）。

> 而且显示它们存在，用户才知道这个自定义的类可以用它们，以及它们的用法。

我赞同这句话说的 “用户 **需要** 知道这个自定义的类可以用它们，以及它们的用法”，但是我不赞同通过直接导出 `__call__` 和 `__repr__` 文档的方式。

比如，PyTorch 的 `nn.Module`， 为什么不让用户直接重写 `__call__`，而是重写 `forward` 呢？我觉得某种意义上就是怕 `__call__`  “古怪”，而且看他对 `forward` 的文档 （ https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.forward ），也不会介绍背后其实最终是通过 `__call__`，而只是说，考虑到注册hook机制，用户应该直接调用 Module 对象：

> Although the recipe for forward pass needs to be defined within this function, one should call the [Module](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module) instance afterwards instead of this since the former takes care of running the registered hooks while the latter silently ignores them.

对 `__repr__` 的文档我觉得就更容易了，写个例子，告知 Graph 对象可以打印就可以了。也就是把 https://github.com/Oneflow-Inc/oneflow/blob/master/python/oneflow/nn/graph/graph.py#L455 这里的 example，挪到 这里 https://github.com/Oneflow-Inc/oneflow/blob/master/python/oneflow/nn/graph/graph.py#L88 

然后加上一些描述，类似 ” Graph 对象是可以直接打印的，输出结果将展示一些必要的调试信息“。 这样我觉得更有利于传递信息。

Reply:
yaochi上面的讨论我不确定。

我现在还是觉得torch那种做法从清晰、专业的角度来说，并不好。

用户要间接的知道可以去直接call和打印一个类。

实际很多人是很晚才知道nn.Module可以打印。我想就是torch这里文档做的不清楚导致的。

Reply:
比如想告诉别人打印一个类，是不是没有必要暴露__repr__接口。

在文档里加多一个代码块，如：
```python
print(simpleModule)

"""
simpleModule has: in_features = xxx, out_features= xxx
"""
```

Reply:
不说repr了，call带有复杂性：

<img width="739" alt="image" src="https://user-images.githubusercontent.com/3468789/155443078-67d479fd-62b3-4b2f-87de-869118250b1d.png">

至于`__call__`和语法糖之间的关系，我们可以假设用户不知道，但是他不知道也没关系，因为上面的使用例子就展示了这个关系。所以这里是没有什么用户负担的。

<img width="863" alt="image" src="https://user-images.githubusercontent.com/3468789/155443436-8dd57ba7-6301-41e3-8ca6-0c63207d4780.png">

torch的call其实也有复杂性，它没有说明出来，比如hook的问题。这是不好的。

nn.Graph的call更特殊，它不是建议最好不要调用build，而是禁止直接调用build，直接调用build实际在执行eager。call里面封装了编译、输出处理的很多隐含操作。

我理解展示call接口，说明更清晰，智力负担可以用例子消除，还是比较合理的。